server/balancer.go

```diff
@@ -0,0 +1,316 @@
+package server
+
+import (
+	"github.com/golang/protobuf/proto"
+	"github.com/juju/errors"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/kvproto/pkg/raftpb"
+)
+
+// Operator is the interface to do some operations.
+type Operator interface {
+	// Do does the operator, if finished then return true.
+	Do(req *pdpb.RegionHeartbeatRequest) (bool, *pdpb.RegionHeartbeatResponse)
+	// Check checks whether the operator has been finished.
+	Check(req *pdpb.RegionHeartbeatRequest) bool
+}
+
+func containPeer(region *metapb.Region, peer *metapb.Peer) bool {
+	for _, p := range region.GetPeers() {
+		if p.GetId() == peer.GetId() {
+			return true
+		}
+	}
+
+	return false
+}
+
+func leaderPeer(region *metapb.Region, storeID uint64) *metapb.Peer {
+	for _, peer := range region.GetPeers() {
+		if peer.GetStoreId() == storeID {
+			return peer
+		}
+	}
+
+	return nil
+}
+
+// BalanceOperator is used to do region balance.
+type BalanceOperator struct {
+	index int
+	ops   []Operator
+}
+
+func newBalanceOperator(oldLeader, newLeader, newPeer *metapb.Peer) *BalanceOperator {
+	leaderTransferOperator := newTransferLeaderOperator(oldLeader, newLeader)
+
+	addPeer := &pdpb.ChangePeer{
+		ChangeType: raftpb.ConfChangeType_AddNode.Enum(),
+		Peer:       newPeer,
+	}
+	addPeerOperator := newChangePeerOperator(addPeer)
+
+	removePeer := &pdpb.ChangePeer{
+		ChangeType: raftpb.ConfChangeType_RemoveNode.Enum(),
+		Peer:       oldLeader,
+	}
+	removePeerOperator := newChangePeerOperator(removePeer)
+
+	return &BalanceOperator{
+		index: 0,
+		ops:   []Operator{leaderTransferOperator, addPeerOperator, removePeerOperator},
+	}
+}
+
+// Check implements Operator.Check interface.
+func (bo *BalanceOperator) Check(req *pdpb.RegionHeartbeatRequest) bool {
+	if bo.index >= len(bo.ops) {
+		return true
+	}
+
+	return false
+}
+
+// Do implements Operator.Do interface.
+func (bo *BalanceOperator) Do(req *pdpb.RegionHeartbeatRequest) (bool, *pdpb.RegionHeartbeatResponse) {
+	// TODO: optimize it to return next operator response directly.
+	if bo.Check(req) {
+		return true, nil
+	}
+
+	finished, res := bo.ops[bo.index].Do(req)
+	if !finished {
+		return false, res
+	}
+
+	bo.index++
+	return false, nil
+}
+
+// ChangePeerOperator is used to do peer change.
+type ChangePeerOperator struct {
+	changePeer *pdpb.ChangePeer
+}
+
+func newChangePeerOperator(changePeer *pdpb.ChangePeer) *ChangePeerOperator {
+	return &ChangePeerOperator{
+		changePeer: changePeer,
+	}
+}
+
+// Check implements Operator.Check interface.
+func (co *ChangePeerOperator) Check(req *pdpb.RegionHeartbeatRequest) bool {
+	region := req.GetRegion()
+	if region == nil {
+		return false
+	}
+
+	leaderPeer := req.GetLeader()
+	if leaderPeer == nil {
+		return false
+	}
+
+	if co.changePeer.GetChangeType() == raftpb.ConfChangeType_AddNode {
```

Comment:
should we check epoch here? 
E,g, region has 1, 2, if we want to add 3, but the report is 1,2, 4, how to do it now?
I think we should do some check to cancel this balance if possible.